python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+                if len(runs) != 1:
+                    raise RuntimeError(f'Got data from runs {runs} with assertSameRun==True')
+
+            for visit in visitList:
+                imageType = butler.queryMetadata('raw', self.config.imageTypeKey, dataId={'visit': visit})[0]
+                imageType = imageType.lower()
+                dataRef.dataId['visit'] = visit
+
+                if imageType == 'flat':  # note different isr tasks
+                    exp = self.isrForFlats.runDataRef(dataRef).exposure
+                    defects = self.findHotAndColdPixels(exp, imageType)
+                    defectListFromFlats.append(defects)
+
+                elif imageType == 'dark':
+                    exp = self.isrForDarks.runDataRef(dataRef).exposure
+                    defects = self.findHotAndColdPixels(exp, imageType)
+                    defectListFromDarks.append(defects)
+
+                else:
+                    raise RuntimeError("Failed on imageType {imageType}. Only flats and darks supported")
+
+                msg = f"Found {len(defects)} defects containing {self._nPixFromDefects(defects)} pixels"
+                msg += f" in visit {visit}"
+                self.log.info(msg)
+
+                if self.config.makePlots:
+                    nSig = self.config.nSigmaBright if imageType == 'flat' else self.config.nSigmaDark
+                    self._plot(dataRef, exp, visit, nSig, defects, imageType)
+
+        self.log.info(f'Combining {len(defectListFromDarks)} defect sets from darks for detector {detNum}')
+        mergedDefectsFromDarks = self._postProcessDefectSets(defectListFromDarks, exp.getDimensions())
+        self.log.info(f'Combining {len(defectListFromFlats)} defect sets from flats for detector {detNum}')
+        mergedDefectsFromFlats = self._postProcessDefectSets(defectListFromFlats, exp.getDimensions())
+
+        self.log.info(f'Combining bright and dark defect sets')
+        brightDarkPostMerge = [mergedDefectsFromDarks, mergedDefectsFromFlats]
+        allDefects = self._postProcessDefectSets(brightDarkPostMerge, exp.getDimensions(), mode='OR')
+
+        self._writeData(dataRef, allDefects)
+
+        self.log.info('Finished finding defects in detector %s' % detNum)
+        return pipeBase.Struct(exitStatus=0)
+
+    @staticmethod
+    def _nPixFromDefects(defects):
+        """TODO: Write docstring
+        """
+        nPix = 0
+        for defect in defects:
+            nPix += defect.getBBox().getArea()
+        return nPix
+
+    def _writeData(self, dataRef, defects):
+        """Write the data out to the defect file.
+
+        TODO: Add parameters
+        """
+        filename = dataRef.getUri(write=True)  # does not guarantee that full path exists
+        dirname = os.path.dirname(filename)
+        if not os.path.exists(dirname):
+            os.makedirs(dirname)
+
+        self.log.info(f'Writing defects to {filename[:-5]} in format: {self.config.writeAs}')
```

Comment:
That's stripping the `.fits` from the end, because we don't know there what format it is, and therefore what the extension will be.

Reply:
Firstly, f-strings should not be used in logging messages since you want delayed substitution. Secondly you should really use `os.path.splitext()` if you want the file name without the extension. That's much safer in the long term.

Reply:
Yes, John pointed out the f-strings, am changing them. Will stick in a `splitext()` here instead too 👍